### PR TITLE
Cleanup localBackup

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 453463ba22e3083dd9e848b3a6089cb0be34ea98 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52837/453463ba22e3083dd9e848b3a6089cb0be34ea98/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/MergeTree/localBackup.cpp [hunk 1]
```diff
@@ -136,10 +133,30 @@ void localBackup(
...
                 disk->copyDirectoryContent(source_path, disk, destination_path);
+                cleanup.success();
+            }
             else
-                localBackupImpl(disk, disk_transaction.get(), source_path, destination_path, make_source_readonly, 0, max_level, copy_instead_of_hardlinks, files_to_copy_intead_of_hardlinks);
+            {
+                std::function<void()> cleaner;
+                if (disk->supportZeroCopyReplication())
+                    /// Note: this code will create garbage on s3. We should always remove `copy_instead_of_hardlinks` files.
+                    /// The third agrument should be a list of excpetions, but (looks like) it is ignored for keep_all_shared_data = true.
```
Comment (serxa) [inline src/Storages/MergeTree/localBackup.cpp:151]:
```suggestion
                    /// The third argument should be a list of exceptions, but (looks like) it is ignored for keep_all_shared_data = true.
```

**Review by serxa — APPROVED** [review]

**Comment by KochetovNicolai** [comment]
Stateless tests (debug) https://github.com/ClickHouse/ClickHouse/pull/52878

**Comment by KochetovNicolai** [comment]
Upgrade check - [a crash in join](https://pastila.nl/?008065e9/0a2d94f2a8f4865624ca5eac89f219f3), is being investigated by @vdimir

**Pull request merged**